feat: Spawn layout evaluation

AdamGS opened the pull request:
very hacky demo

AdamGS committed:
.

AdamGS committed:
.

AdamGS commented on vortex-layout/src/scan/task.rs:
```diff
@@ -36,17 +35,14 @@ impl ScanTask {
-#[async_trait]
 impl TaskExecutor for InlineTaskExecutor {
-    async fn execute(&self, array: &Array, tasks: &[ScanTask]) -> VortexResult<Array> {
+    fn execute(&self, array: &Array, tasks: &[ScanTask]) -> VortexResult<Array> {
```
this is a flyby but seems like it can be blocking now? potentially not even needed but I'm not sure what are your plans for it

AdamGS committed:
.

AdamGS committed:
enable feature

AdamGS committed:
.

AdamGS commented:
made another round of changes, I think this interface is more general (and works better for the non-specific runtime case), but `spawn` might not be the best name here, as for the inline variant awaiting is necessary to make progress. It fits our current usecase which is good enough for now IMO

AdamGS committed:
Merge branch 'develop' into adamg/spawn-evaluate

AdamGS commented:
I have an idea for the non-specific-runtime variant, coding it now

AdamGS committed:
nicer version

AdamGS commented on vortex-layout/src/scan/executor/threads.rs:
```diff
@@ -0,0 +1,96 @@
+    result: oneshot::Sender<R>,
+}
+
+impl<F, R> Task for ExecutorTask<F, R>
```
definitely not taken from @a10y 😇

AdamGS committed:
.

AdamGS committed:
.

AdamGS committed:
.

AdamGS committed:
.

AdamGS committed:
some work, need to unifiy to one spawn

AdamGS committed:
.

AdamGS committed:
Merge branch 'develop' into adamg/spawn-evaluate

AdamGS committed:
.

AdamGS commented on vortex-layout/src/scan/mod.rs:
```diff
@@ -236,6 +242,7 @@ pub struct Scan<D> {
     filter: Option<ExprRef>,
     row_masks: Vec<RowMask>,
     canonicalize: bool,
+    #[allow(dead_code)] //TODO(adam): bake this into the executors?
```
key point

AdamGS commented on vortex-layout/src/scan/mod.rs:
```diff
@@ -272,52 +276,55 @@ impl<D: ScanDriver> Scan<D> {
+
+                Ok(processed_array)
+            }
+            .map(|v| v.transpose())
```
this is so ugly, but I couldn't figure out any other way to do that

gatesn commented on vortex-layout/src/scan/mod.rs:
```diff
@@ -68,7 +70,9 @@ impl<D: ScanDriver> ScanBuilder<D> {
             split_by: SplitBy::Layout,
             canonicalize: false,
-            concurrency: 32,
+            concurrency: std::thread::available_parallelism()
```
I don't know that this should default to nthreads, this is "concurrency" and not "parallelism", the number of row splits to make progress on concurrently. Basically... we want as many as possible provided the data fits in some reasonable memory, since the more concurrency, the more visibility into upcoming I/O, and therefore the more coalescing.

AdamGS replied:
yeah, it was an idea but we probably do want to oversubscribe the tokio runtime if we can and buffer by a different metric

AdamGS committed:
VortexResult::flatten -> unnest (#2361)

Didn't actually check that in #2360, apparently the compiler doesn't
like `flatten` becuase it exists in an unstable feature.

This PR is dedicated to @danking as the chief
[haskeller](https://en.wiktionary.org/wiki/Haskeller).

AdamGS committed:
.

AdamGS committed:
.

AdamGS committed:
Merge branch 'develop' into adamg/spawn-evaluate

AdamGS commented:
seems like I lost the performance benefit somewhere along the way. It might be due to another change but I don't think we merged anything major since the first run

```
> git log 060d7137c..489262d5e7f44bec091e12141133d506e879e4e2 --oneline
489262d5e (HEAD -> develop, origin/develop, origin/HEAD) VortexResult::flatten -> unnest (#2361)
3763a26a5 add ext trait and function to flatten nested VortexResult (#2360)
68e18328d fix: bring back bench test-harness flag (#2357)
c3ce79b26 feat: allow benchmarking against remote files (#2297)
4d97e227b chore: avoid AWS machines for now in benchmark actions (#2354)
e31512d17 bench: enable divan codspeed compatibility layer (#2351)
7e96d79e5 Dict encoder can encode multiple arrays with same dictionary again (#2353)
978ae150b `take_into` compute function (#2344)
bff9fc2c6 chore(deps): update rust crate criterion to v2.8.0 (#2349)
74fac0671 add max coalesced read performance hint (#2346)
081593ebe fix: Chunked compressor doesn't fail if there's uncompressible chunk (#2347)
```

AdamGS committed:
.

AdamGS committed:
.

AdamGS committed:
go back to higher default concurrency, not sure how to pick that number

AdamGS commented:
Perf change here isn't as dramatic, but definitely consistent, so seems like this is a good direction for now, running benchmarks on one last change before this is ready for review.

AdamGS committed:
.

AdamGS committed:
.

AdamGS commented on vortex-layout/src/scan/mod.rs:
```diff
@@ -312,17 +311,14 @@ impl<D: ScanDriver> Scan<D> {
+                async move { task_executor.spawn(processing_task) }
+            })
+            .try_buffered(self.concurrency)
+            .filter_map(|v| async move { v.unnest().transpose() });
```
both @robert3005 and I suspect that this is what we actually want here, defering the `await` on the spawned task as much as possible while still buffering the `spawn` calls.

AdamGS committed:
Merge branch 'develop' into adamg/spawn-evaluate

AdamGS committed:
.

AdamGS committed:
Merge branch 'develop' into adamg/spawn-evaluate

AdamGS committed:
.

AdamGS committed:
.

AdamGS committed:
Merge branch 'develop' into adamg/spawn-evaluate

AdamGS committed:
Merge branch 'develop' into adamg/spawn-evaluate

AdamGS commented on vortex-layout/src/scan/executor/threads.rs:
```diff
@@ -0,0 +1,96 @@
+    }
+}
+
+struct Inner {
```
This structure is needed to achieve the behavior we want, where arbitrary futures can be spawned on this runtime. I've tried to structure it differently but I honestly think this is pretty nice, and we can definitely try and improve that later.

AdamGS committed:
cleanup ScanExecutor, move some tests to rstest

AdamGS committed:
Merge branch 'develop' into adamg/spawn-evaluate

AdamGS committed:
Merge branch 'develop' into adamg/spawn-evaluate

AdamGS committed:
typo

AdamGS commented:
All benchmarks look good, not much difference on S3 (maybe slightly better) and much better on disk (especially for TPC-H).

gatesn committed:
Merge branch 'develop' into adamg/spawn-evaluate

gatesn commented on vortex-layout/src/layouts/flat/eval_expr.rs:
```diff
@@ -35,7 +35,11 @@ impl ExprEvaluator for FlatReader {
-        self.executor().evaluate(&array, &tasks).await
+        let mut array = array;
+        for task in tasks {
+            array = task.execute(&array)?;
```
If we're just going to run these inline, let's get rid of ScanTask entirely

AdamGS replied:
done

gatesn commented on vortex-layout/src/scan/executor/mod.rs:
```diff
@@ -0,0 +1,49 @@
+use vortex_error::VortexResult;
+
+pub trait Executor {
+    /// Spawns a future to run on a different runtime. The shouldn't be polled to make progress.
```
Fix up doc

AdamGS replied:
LMK what you think of the new one

gatesn commented on vortex-layout/src/scan/executor/mod.rs:
```diff
@@ -0,0 +1,49 @@
+
+pub trait Executor {
+    /// Spawns a future to run on a different runtime. The shouldn't be polled to make progress.
+    fn spawn<F>(&self, f: F) -> VortexResult<BoxFuture<'static, VortexResult<F::Output>>>
```
Does this need to return a result? Can it just be `BoxFuture<'static, VortexResult<..>>`? Since the impl can always return `ready(Err(...))` if it needs to.

AdamGS replied:
I don' think we want to return `ready(Err(...))` here, but if we're ok with panicking in cases where we can't submit work/tasks start getting cancelled, we can get the nicer function signature.

gatesn commented on vortex-layout/src/scan/executor/mod.rs:
```diff
@@ -0,0 +1,49 @@
+
+/// Generic wrapper around different async runtimes. Used to spawn async tasks to run in the background, concurrently with other tasks.
+#[derive(Clone)]
+pub enum TaskExecutor {
```
I wonder what the performance hit is to just have a `dyn Executor`? Seems like a lot of code here for basically `|f| tokio::spawn(f)`

AdamGS replied:
IDK if we can have `dyn Executor` here without a major change due to the generics

gatesn replied:
You would have to take / return BoxFuture. You already return one

AdamGS replied:
I don't think its that much code (probably less than what happens in `Arc`), but in order to make it dyn-compatible we'll also have to move the generic bound from the function to the trait itself ([example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0206bd498372471b795fe25f0ee1d5a1)), which makes this thing less useful in other cases IMO.

AdamGS committed:
CR comments

AdamGS committed:
.

gatesn commented on vortex-layout/src/scan/executor/mod.rs:
```diff
@@ -0,0 +1,50 @@
+    Tokio(TokioExecutor),
+}
+
+pub enum TaskExecutorMode {
```
This is never used

gatesn commented on vortex-layout/Cargo.toml:
```diff
@@ -21,18 +21,20 @@ bit-vec = { workspace = true }
 exponential-decay-histogram = { workspace = true }
 flatbuffers = { workspace = true }
-futures = { workspace = true, features = ["alloc"] }
+flume = { workspace = true }
```
Can this be optional behind a `threads` feature?

AdamGS replied:
so WDYT should be the default behavior?

AdamGS committed:
remove unused code
